src/utils/actions.js

```diff
@@ -0,0 +1,325 @@
+//      trust services
+//      database
+//  reportingBundle: the entire bundle to be submitted
+//  contentBundle: just the content section of the report bundle
+//  next: place to put arbitrary objects specifically for the next step?
+//  prev: where arbitrary objects from the previous set should go?
+//  flags: various boolean or enum flags about the status of the report.
+//         is it encrypted y/n, is it identifiable or de/pseudo/anonymized, etc
+//  cancelToken: indicates the job should be cancelled (from a process external
+//         to the job itself -- this tells the job to cease processing)
+//  exitStatus: if set, terminates the job and marks it with the given status
+//         (for example, if a given patient doesn't meet reporting criteria, don't keep reporting)
+
+// we assume the context updates and is passed to functions
+// in turn.  I'm also assuming the context is unique
+// to the current action, since we'd be cycling through actions
+// anyway.
+const Fhir = require('fhir').Fhir;
+const FhirPath = require('fhir').FhirPath;
+const fhir = new Fhir();
+const baseIgActions = {
+  'check-trigger-codes': context => {
+    const action = context.action;
+    const resources = context.records;
+
+    const boolMap = action.action[0].condition.map(condition => {
+      if (condition.expression) {
+        const expression = condition.expression;
+        // check the records for a trigger code
+        return resources.some(resource => {
+          //might have to check that this is non-empty here
+          return evaluateExpression(expression, resource);
+        });
+      } else {
+        // default true for non-applicability conditions
+        return true;
+      }
+    });
+    // boolMap should be all-true (all conditions met)
+    const triggered = boolMap.every(entry => {
+      return entry === true;
+    });
+    context.flags['triggered'] = triggered;
+  },
+  'check-participant-registration': context => {
+    // TODO: how do we check the participant registration?
+    // Through the source client?
+    return context;
+  },
+  'create-report': context => {
+    // the output might define a profile
+    let contentBundle = createBundle(context.records);
+    if (context.action.output) {
+      const output = context.action.output[0];
+      if (output.profile) contentBundle = validateProfile(contentBundle, output.profile[0]);
+    }
+    const header = makeHeader(context);
+    context.contentBundle = contentBundle;
+    context.reportingBundle = createBundle([header, ...contentBundle.entry]);
+  },
+  'validate-report': context => {
+    // check if fhir is well-formed
+    const validation = fhir.validate(context.reportingBundle);
+    context.flags['valid'] = validation.valid;
+  },
+  'submit-report': async context => {
+    await context.client.dest
+      .submit(context.reportingBundle)
+      .then(
+        result => {
+          // might want to check for a 200?
+          if (result.status === 200) {
```

Comment:
So our implementation definitely returns 200 on success but the $process-message docs allow for 202 Accepted to be returned. Since we're not using the response data here is it worth checking for 202 as well?